src/backend/opencl/Array.cpp

```diff
@@ -228,30 +228,66 @@ namespace opencl
                 out.eval();
             } else {
+
                 size_t alloc_bytes, alloc_buffers;
                 size_t lock_bytes, lock_buffers;
 
                 deviceMemoryInfo(&alloc_bytes, &alloc_buffers,
                                  &lock_bytes, &lock_buffers);
 
-                if (lock_bytes > getMaxBytes() ||
-                    lock_buffers > getMaxBuffers()) {
+                bool isBufferLimit =
+                    lock_bytes > getMaxBytes() ||
+                    lock_buffers > getMaxBuffers();
+
+
+                bool isNvidia = getActivePlatform() == AFCL_PLATFORM_NVIDIA;
+                // We eval in the following cases.
+                // 1. Too many bytes are locked up by JIT causing memory pressure.
+                // Too many bytes is assumed to be half of all bytes allocated so far.
+                // 2. Too many buffers in a nonlinear kernel cause param space overflow.
+                // Too many buffers comes out to be about 48 (49 including output).
+                // Too many buffers can occur in a tree of size 24 in the worst case scenario.
+                // This error only happens on nvidia devices.
+                // TODO: Find better solution than the following emperical solution.
+                bool isParamLimit = (isNvidia && node->getHeight() > 24);
+                if (isParamLimit || isBufferLimit) {
 
-                    unsigned length =0, buf_count = 0, bytes = 0;
                     Node *n = node.get();
-                    JIT::Node_map_t nodes_map;
-                    std::vector<JIT::Node *> full_nodes;
-                    std::vector<JIT::Node_ids> full_ids;
+
+                    // Use thread local to reuse the memory every time you are here.
+                    thread_local JIT::Node_map_t nodes_map;
+                    thread_local std::vector<Node *> full_nodes;
+                    thread_local std::vector<JIT::Node_ids> full_ids;
+
+                    // Reserve some memory
+                    if (nodes_map.size() == 0) {
+                        nodes_map.reserve(1024);
+                        full_nodes.reserve(1024);
+                        full_ids.reserve(1024);
+                    }
+
                     n->getNodesMap(nodes_map, full_nodes, full_ids);
 
+                    unsigned length = 0, buf_count = 0, bytes = 0;
+                    bool is_linear = true;
+                    dim_t dims_[] = {dims[0], dims[1], dims[2], dims[3]};
                     for(auto &jit_node : full_nodes) {
                         jit_node->getInfo(length, buf_count, bytes);
+                        is_linear &= jit_node->isLinear(dims_);
                     }
 
-                    if (2 * bytes > lock_bytes) {
+                    // Clear the vectors so you dont hold on to any shared pointers
+                    nodes_map.clear();
+                    full_nodes.clear();
+                    full_ids.clear();
```

Comment:
Aren't you just referencing pointers with these variables? You shouldn't need to do this.

Reply:
`Node_ptr` may contain a shared pointer.

Reply:
Now that I am looking at the getNodesMap function you might still need this because you are doing a push_back in there. I think you just need to update the comment.